### ISSUE TITLE
Scatter data fetching from difference component

### ISSUE THREAD

**Issue by vanhtuan0409** [facebook/flux#332]
Currently I have 3 component (A, B and C) need to use the same kind of data from server
All of 3 component can listen to Store change and get data from Store. But the problem is I dont know where to place the data fetching action.
If all 3 components do the data fetching action, It will cause the same data to be fetched multiple time and component will re-render many time when data received
If I place the data fetching action in a top level component and these 3 components only listen to store change, it will be hard to reuse these 3 components.
In summary, should I fetch the data from server and listen to Store change in the same component (to make component easier to reuse) or should I fetch the data in a top component and listen to Store change in lower level component

**Comment by kyldvs** [comment]
I tend to fire the actions from the store where data is being requested. In order to accomplish what you want you will need to have a sense of "status" of a particular piece of data in your store, not just the raw data.

```js
import {ReduceStore} from 'flux/utils';

type Status<T> = {
  status: 'not_loaded' | 'loading' | 'loaded',
  data?: T,
}

class FooStore extends ReduceStore<Status<string>> {
  getInitialState() {
    return {status: 'not_loaded'};
  }

  reduce(state: Status<string>, action: Object) {
    switch (action.type) {
      case 'start_loading':
        if (state.status === 'loading') {
          return state; // we don't want to generate many requests if we are already loading
        } else {
          FooDataManager.loadFoo();
          return {status: 'loading'};
        }

      case 'foo_loaded':
        return {
          status: 'loaded',
          data: action.foo.name,
        };

      case 'foo_load_error'
        return {status: 'not_loaded'}; // Or create an "error" status and put the error in data
    }
  }

  // Make sure to override this so that we can transition to loading when we need to. Alternatively
  // you could check at every call site to getState() and do the exact same thing instead
  getState() {
    const state = super.getState();
    if (state.status === 'not_loaded') {
      FooDispatcher.dispatch({type: 'start_loading'});
    }
    return state;
  }
}

// We centralize all network requests in a "data manager" or "web api utils"
// call it whatever, just keep them in one place
const FooDataManager = {
  loadFoo() {
    xhr({uri: 'http://api.mywebsite.com/foo'}, (err, resp, body) => {
      if (err) {
        FooDispatcher.dispatch({type: 'foo_load_error'});
        console.error(err); // do whatever you want with the error
      };
      const data = JSON.parse(body); // may want to verify resp code
      FooDispatcher.dispatch({
        type: 'foo_loaded',
        foo: data,
      });
    });
  }
}
```

**Comment by vanhtuan0409** [comment]
I thought we should not fetch data in Store
The data should be fetch via action from View if I recall correctly
But in general I understand the concept of status
One more question is that if status is 'foo_loaded' so we can get data directly from Store and do not fire Action to fetch data. But after a while, the data is out-of-date but still in the Store. How we can refresh the data after a while
I know one solution is using real time Websocket to notify when the data is changed from the server. Other than that is there any work around ?

**Comment by kyldvs** [comment]
That is basically cache invalidation problem :) you can do a simple TTL, and after 5 minutes or so invalidate your data, or the web socket solution you mentioned. After that just fire a new action when you want to clear the data.

It's fine to issue a request from your store, the one critical thing is that the result of that request re-enters the flux system via a separate action. If you do this it would be really bad:

```js
// Counter example, never do this!
class Store {
  onDispatch(action) {
    switch (action.type) {
      case 'start_loading'
        xhr({uri: ...}, (err, resp, body) => {
          this._state = JSON.parse(body); // Don't do this! You are mutating the store outside of a dispatch
        });
    }    
  }
}
```

It's pretty much impossible to do the above though when using reducers though

**Comment by pulse14** [comment]
@kyldvs Am i right that on your FooStore example there is a lot of chance of dispathing in a middle of another dispatching ? 

I know from the conf you track almost all loading state at FB. But the AdsManager seems to don't track loading with actions but stores directly mutate states. Since these stores are mainly used for initialization I think there is nothing wrong with that ? You just mutate the last state because you know you can't give it to the components since this is not what they ask.

is it still a good practice ?

**Comment by kyldvs** [comment]
No there should be no chance of dispatching in the middle of a dispatch otherwise your application will crash. In my simplified example yeah it could be possible if you are calling `getState()` within another store. In practice I think we do something like this (in some base class/util so we don't keep writing it):

```
  getState() {
    const state = super.getState();
    if (state.status === 'not_loaded') {
      if (FooDispatcher.isDispatching()) {
        setImmediate(() => FooDispatcher.dispatch({type: 'start_loading'}));
      } else {
        FooDispatcher.dispatch({type: 'start_loading'});
      }
    }
    return state;
  }
```

Yes used to we would directly mutate the state of a store from `not_loaded` to `loading` without emitting a change. It generally works and is a bit simpler, but strictly speaking it breaks some of the core principles of the Flux architecture so we moved away from that. In particular you should always emit a change when your state changes and you must only emit a change during a dispatch. In the old system it was impossible to maintain these two conditions.

--

No we do not have any plans to add an after dispatch callback. We played around with it for a bit, but it was not ever used and generally complicated the system.

**Comment by vanhtuan0409** [comment]
@kyldvs One more question is in FB, how do you manage the loading state of entity
For example, I have BookStore to manage all the book in library. In my application, sometime I need to fetch data for 1 book only, but sometime I need to get data for all available book
How do you design the state to manage both the single item fetching and group item fetching

**Comment by kyldvs** [comment]
https://github.com/facebook/flux/tree/master/examples includes how we handle async actions now, hopefully that will provide enough guidance to consider this issue resolved.
